pkg/reconciler/taskrun/validate_taskrun_test.go

```diff
@@ -104,142 +106,385 @@ func TestValidateResolvedTask_ValidParams(t *testing.T) {
+				"key2": "val2",
+				"key3": "val3",
+			}),
 		}},
+		matrix: &v1beta1.Matrix{},
+	}, {
+		name: "extra-param-matrix",
+		task: v1beta1.Task{
+			ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+			Spec: v1beta1.TaskSpec{
+				Params: []v1beta1.ParamSpec{{
+					Name: "include",
+					Type: v1beta1.ParamTypeString,
+				}},
+			},
+		},
+		params: v1beta1.Params{{}},
 		matrix: &v1beta1.Matrix{
 			Params: []v1beta1.Param{{
-				Name:  "barfoo",
-				Value: *v1beta1.NewStructuredValues("bar", "foo"),
+				Name:  "extraArr",
+				Value: *v1beta1.NewStructuredValues("extra", "arr"),
+			}},
+			Include: []v1beta1.IncludeParams{{
+				Name: "build-1",
+				Params: v1beta1.Params{{
+					Name: "include", Value: *v1beta1.NewStructuredValues("string"),
+				}},
 			}},
 		},
+	}}
+	for _, tc := range tcs {
+		rtr := &resources.ResolvedTask{
+			TaskSpec: &tc.task.Spec,
+		}
+		t.Run(tc.name, func(t *testing.T) {
+			if err := ValidateResolvedTask(ctx, tc.params, tc.matrix, rtr); err != nil {
+				t.Errorf("Did not expect to see error when validating TaskRun with correct params but saw %v", err)
+			}
+		})
+	}
+}
+
+func TestValidateResolvedTask_InvalidParams(t *testing.T) {
+	ctx := context.Background()
+	tcs := []struct {
+		name    string
+		task    v1beta1.Task
+		params  []v1beta1.Param
+		matrix  *v1beta1.Matrix
+		wantErr string
+	}{{
+		name: "missing-params-string",
+		task: v1beta1.Task{
+			ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+			Spec: v1beta1.TaskSpec{
+				Params: []v1beta1.ParamSpec{{
+					Name: "foo",
+					Type: v1beta1.ParamTypeString,
+				}},
+			},
+		},
+		params: v1beta1.Params{{
+			Name:  "missing",
+			Value: *v1beta1.NewStructuredValues("somethingfun"),
+		}},
+		matrix:  &v1beta1.Matrix{},
+		wantErr: "invalid input params for task : missing values for these params which have no default values: [foo]",
 	}, {
-		name: "invalid-type-in-params",
-		rtr: &resources.ResolvedTask{
-			TaskSpec: &task.Spec,
+		name: "missing-params-arr",
+		task: v1beta1.Task{
+			ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+			Spec: v1beta1.TaskSpec{
+				Params: []v1beta1.ParamSpec{{
+					Name: "foo",
+					Type: v1beta1.ParamTypeArray,
+				}},
+			},
+		},
+		params: v1beta1.Params{{
+			Name:  "missing",
+			Value: *v1beta1.NewStructuredValues("array", "param"),
+		}},
+		matrix:  &v1beta1.Matrix{},
+		wantErr: "invalid input params for task : missing values for these params which have no default values: [foo]",
+	}, {
+		name: "invalid-string-param",
+		task: v1beta1.Task{
+			ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+			Spec: v1beta1.TaskSpec{
+				Params: []v1beta1.ParamSpec{{
+					Name: "foo",
+					Type: v1beta1.ParamTypeString,
+				}},
+			},
+		},
+		params: v1beta1.Params{{
+			Name:  "foo",
+			Value: *v1beta1.NewStructuredValues("array", "param"),
+		}},
+		matrix:  &v1beta1.Matrix{},
+		wantErr: "invalid input params for task : param types don't match the user-specified type: [foo]",
```

Comment:
Suggest including more info in this error message comparing the type received to the type specified; see https://github.com/tektoncd/pipeline/issues/6365#issuecomment-1481243995 for more detail on why.
Although this might be better in a separate PR